### PR TITLE
feat(trace): span streaming, filtering, and exporting

### LINKED ISSUE #1198: [traces][gq] expose a query str param to the spans connection
As a user, I'd like to search for spans, traces etc that match certain criteria like 

`attributes.llm.name == "gpt-3.5-turbo" & total_token_count > 50`

The idea here is that we can leverage https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html to make a very powerful search experience.

The UI would then have a certain level of flexibility to construct these queries.

### LINKED ISSUE #1234: [ENHANCEMENT] span streaming
Based on stakeholder feedback on 8/25:
- As a user, I don't want to interrupt my workflow to start and stop phoenix in order to load the traces.
- As a user, I want to be able to run multiple workflows/notebooks at the same time and see their traces in the same view.

### LINKED ISSUE #1242: [evals] Export trace data into dataframe
- [ ] add dataframe exporter

### PR DESCRIPTION
**Pull request by RogerHYang**
resolves #1234
resolves #1242
resolves #1198

# Streaming

https://github.com/Arize-ai/phoenix/assets/80478925/16bc0d79-0131-4a3a-9ef7-e295e55057ba

# Filtering and Exporting

https://github.com/Arize-ai/phoenix/assets/80478925/71dcd9fb-105b-4272-8b80-9bed46c17e0d

### CODE REVIEW

File: src/phoenix/server/api/types/Span.py [hunk 1]
```diff
@@ -1,36 +1,26 @@
...
 from enum import Enum
 from typing import Any, DefaultDict, List, Mapping, Optional, cast
 
 import strawberry
-from pandas import Series
 from strawberry import ID
 from strawberry.types import Info
 
-import phoenix.trace.semantic_conventions as sc
+import phoenix.trace.schemas as s
```
Comment (axiomofjoy) [inline src/phoenix/server/api/types/Span.py:11]: More semantic package names are easier to understand IMO (excepting common abbreviations like `pd` and `np` and `tf`).

File: src/phoenix/core/traces.py [hunk 2]
```diff
@@ -1,108 +1,288 @@
...
+        # Putting `None` as the sentinel value for queue termination.
+        weakref.finalize(self, self._queue.put, None)
+        for span in spans or ():
+            self._queue.put(span)
+        self._lock = Lock()
+        self._spans: Dict[SpanID, ReadableSpan] = {}
+        self._parent_span_ids: Dict[SpanID, ParentSpanID] = {}
+        self._traces: Dict[TraceID, List[SpanID]] = defaultdict(list)
+        self._child_span_ids: DefaultDict[SpanID, List[ChildSpanID]] = defaultdict(list)
+        self._orphan_spans: DefaultDict[ParentSpanID, List[Span]] = defaultdict(list)
```
Comment (axiomofjoy) [inline src/phoenix/core/traces.py:93]: Nit: These names sound more like lists than dictionaries.

File: src/phoenix/core/traces.py [hunk 3]
```diff
@@ -1,108 +1,288 @@
...
+            )
+            new_span[cumulative_attribute_name] = difference + existing_cumulative_value
+            self._add_value_to_span_ancestors(
+                difference,
+                span_id,
+                cumulative_attribute_name,
+            )
+        # Process previously orphan spans if any
+        for orphan_span in self._orphan_spans[span_id]:
+            self._process_span(orphan_span)
```
Comment (axiomofjoy) [inline src/phoenix/core/traces.py:267]: What situation do we wind up with orphan spans?

File: src/phoenix/core/traces.py [hunk 4]
```diff
@@ -1,108 +1,288 @@
...
+            return self._spans.get(key)
+
+    def _consume_spans(self) -> None:
+        while True:
+            if not (span := self._queue.get()):
+                return
+            with self._lock:
+                self._process_span(span)
+
+    def _process_span(self, span: Span) -> None:
```
Comment (axiomofjoy) [inline src/phoenix/core/traces.py:213]: Docstring would help here, or breaking into smaller parts that are easier to understand. There's a lot going on in this one.

**Comment by axiomofjoy** [comment]
Looks good to me. Let's get Mikyo's eyes on it as well.

File: src/phoenix/core/traces.py (hunk 3 above)
Comment (RogerHYang) [reply src/phoenix/core/traces.py:267]: In general, network delivery has no guarantee on arrival order (even though things are sent in order).

File: src/phoenix/core/traces.py [hunk 5]
```diff
@@ -1,108 +1,288 @@
...
+            datetime,
+            self._min_start_time,
+        )
+        stop_time = stop_time or (
+            cast(
+                datetime,
+                self._max_start_time,
+            )
+            + timedelta(minutes=1)
+        )
```
[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/core/traces.py:140]: stop_time is exclusive, but max_start_time belongs to an actual record, so to include that record we need to pad the max time with something extra

File: src/phoenix/core/traces.py (hunk 2 above)
Comment (RogerHYang) [reply src/phoenix/core/traces.py:93]: Dicts and lists are very similar in that lists are just dicts with integer keys.

File: src/phoenix/core/traces.py [hunk 6]
```diff
@@ -1,108 +1,288 @@
...
-        df[LATENCY_MS] = (df[END_TIME] - df[START_TIME]).dt.total_seconds() * 1000
-        parent_span_ids = df.loc[:, PARENT_ID]
-        for column, cumulative_column in (
+    def __init__(
+        self,
+        spans: Optional[Iterable[Span]] = None,
+    ) -> None:
+        self._queue: "SimpleQueue[Optional[Span]]" = SimpleQueue()
+        # Putting `None` as the sentinel value for queue termination.
+        weakref.finalize(self, self._queue.put, None)
```
[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/core/traces.py:85]: `weakref.finalize` carries out clean up operations when the object (in this case `self`) is being garbage collected. The clean up operation here is to make sure the thread terminates by putting a sentinel value in the queue.

File: src/phoenix/core/traces.py [hunk 7]
```diff
@@ -1,108 +1,288 @@
...
-        df = df.set_index(SPAN_ID, drop=False)
-        df[LATENCY_MS] = (df[END_TIME] - df[START_TIME]).dt.total_seconds() * 1000
-        parent_span_ids = df.loc[:, PARENT_ID]
-        for column, cumulative_column in (
+    def __init__(
+        self,
+        spans: Optional[Iterable[Span]] = None,
+    ) -> None:
+        self._queue: "SimpleQueue[Optional[Span]]" = SimpleQueue()
+        # Putting `None` as the sentinel value for queue termination.
```
[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/core/traces.py:84]: A queue can be empty when it's waiting for more stuff. A sentinel value is needed to tell it to stop waiting.

File: src/phoenix/core/traces.py [hunk 8]
```diff
@@ -1,108 +1,288 @@
...
+        if key.startswith(ATTRIBUTE_PREFIX):
+            return self.__wrapped__.attributes.get(
+                key[len(ATTRIBUTE_PREFIX) :],
+            )
+        return getattr(self.__wrapped__, key, None)
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        if not key.startswith(COMPUTED_NUMERIC_PREFIX):
+            raise KeyError(f"{key} is not a computed value")
+        self._self_computed_numeric_values[key] = value
```
[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/core/traces.py:71]: This name is [borrowed](https://github.com/open-telemetry/opentelemetry-python/blob/9ced357e9ffe32eabe03142848d0c206d026d441/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L346-L347) from OTEL. In our case we allow some write access.

File: src/phoenix/server/api/types/Span.py (hunk 1 above)
Comment (RogerHYang) [reply src/phoenix/server/api/types/Span.py:11]: In this case it's just for resolving naming conflicts between the graphql module and our core module. It suffices to keep it short because it serves a very narrow purpose: i.e. as a prefix for disambiguation. The alternative is to rename everything with a `Core` prefix, which looks a bit strange to me (wait, what is a CoreSpan? never seen that before...oh I see).

option 1: Span vs. s.Span (import schema as s)
option 2: Span vs. CoreSpan (from schema import Span as CoreSpan)

File: src/phoenix/session/session.py [hunk 9]
```diff
@@ -85,7 +85,10 @@ def __init__(
             else None
         )
 
-        self.traces = Traces(trace_dataset.dataframe) if trace_dataset is not None else None
+        self.traces = Traces()
+        if trace_dataset:
+            for span in trace_dataset.to_spans():
+                self.traces.put(span)
```
[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/session/session.py:91]: we could, but it's not adding a lot of value and i don't want to create two ways (and more code) for doing the same thing.

File: src/phoenix/server/main.py [hunk 10]
```diff
@@ -102,10 +108,17 @@ def _get_pid_file() -> str:
...
-        traces = Traces(traces_ds.dataframe)
+        for span in map(
+            json_string_to_span,
+            _download_traces_fixture(
+                _get_trace_fixture_by_name(
+                    trace_dataset_name,
+                ),
+            ),
+        ):
+            traces.put(span)
```
[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/server/main.py:121]: we could, but it's not adding a lot of value and i don't want to create two ways (and more code) for doing the same thing.

File: src/phoenix/trace/schemas.py [hunk 11]
```diff
@@ -12,6 +12,10 @@ class SpanStatusCode(Enum):
     OK = "OK"
     ERROR = "ERROR"
 
+    @classmethod
+    def _missing_(cls, v: Any) -> Optional["SpanStatusCode"]:
+        return None if v else cls.UNSET
+
```
[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/trace/schemas.py:18]: In case the status code is missing, i.e. `span.get("status_code") is None`, it will default to `UNSET` when we do `SpanStatus(span.get("status_code"))`. It makes the code shorter. But it will still error out if we do `SpanStatus("UNK")` because "UNK" is not valid.

File: src/phoenix/trace/tracer.py [hunk 12]
```diff
@@ -13,6 +21,51 @@
...
+
+class HttpExporter:
+    def __init__(self, port: int = PORT) -> None:
+        self._url = f"http://localhost:{port}/v1/spans"
+        self._session = Session()
+        weakref.finalize(self, self._session.close)
+        self._session.headers.update({"content-type": "application/octet-stream"})
+        self._queue: "SimpleQueue[Optional[Span]]" = SimpleQueue()
+        # Putting `None` as the sentinel value for queue termination.
+        weakref.finalize(self, self._queue.put, None)
```
[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/trace/tracer.py:48]: see the other reply for the same question

File: src/phoenix/core/traces.py (hunk 4 above)
Comment (RogerHYang) [reply src/phoenix/core/traces.py:211]: We can refactor this in the future when we have a better understanding. Overhead should be minimized for the first version of this.

File: src/phoenix/core/traces.py [hunk 13]
```diff
@@ -1,108 +1,288 @@
...
+
+    @property
+    def max_start_time(self) -> Optional[datetime]:
+        with self._lock:
+            return self._max_start_time
+
+    def __len__(self) -> int:
+        """Total number of traces"""
+        with self._lock:
+            return len(self._start_time_sorted_root_span_ids)
```
[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/core/traces.py:198]: i'll remove this for now since it's not yet being used

File: src/phoenix/core/traces.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (axiomofjoy) [reply src/phoenix/core/traces.py:71]: Still seems to me like a bit of a misnomer since that class is purely read-only.

[the remark this answers is not in the document]
Comment (axiomofjoy) [reply src/phoenix/core/traces.py:71]: Just a nit.

[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/core/traces.py:71]: Well, readable doesn't mean read only: it's a misnomer for them, not for us. haha

File: src/phoenix/trace/schemas.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (axiomofjoy) [reply src/phoenix/trace/schemas.py:18]: Doesn't this need to be a dunder?

File: src/phoenix/session/session.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (axiomofjoy) [reply src/phoenix/session/session.py:91]: It looks like the `Traces` class is already implemented to accept a list of spans to the initializer.

File: src/phoenix/server/main.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (axiomofjoy) [reply src/phoenix/server/main.py:121]: Same as above.

File: src/phoenix/trace/schemas.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/trace/schemas.py:18]: no dunder, just [`_missing_`](https://docs.python.org/3/library/enum.html#enum.Enum._missing_)

File: src/phoenix/session/session.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/session/session.py:91]: Oh you're right. I was going to take it out but forgot..I'll just leave it then, since it's probably an intuitive expectation to have it as an option at `__init__` after all.

File: src/phoenix/core/traces.py [hunk 14]
```diff
@@ -1,108 +1,267 @@
...
+        # Process previously orphaned spans, if any.
+        for orphan_span in self._orphan_spans[span_id]:
+            self._process_span(orphan_span)
+
+    def _add_value_to_span_ancestors(
+        self,
+        value: float,
+        span_id: SpanID,
+        cumulative_attribute_name: str,
+    ) -> None:
```
Comment (mikeldking) [inline src/phoenix/core/traces.py:262]: the function signature is a bit complex and in a strange order?
```suggestion
    def _add_value_to_span_ancestors(
        self,
        span_id: SpanID,
        attribute_name: str,
        value: float,
    ) -> None:
```

File: src/phoenix/datetime_utils.py [hunk 15]
```diff
@@ -70,3 +70,16 @@ def floor_to_minute(dt: datetime) -> datetime:
 
 
 MINUTE_DATETIME_FORMAT = "%Y-%m-%dT%H:%M:00%z"
+
+
+def time_range(
```
Comment (mikeldking) [inline src/phoenix/datetime_utils.py:75]: this is not a standard time range, might name it a bit different, like `right_open_time_range` - it will read a bit cleaner at the callsite. otherwise it could be confusing why this is event needed.

File: src/phoenix/experimental/callbacks/langchain_tracer.py [hunk 16]
```diff
@@ -155,13 +156,17 @@ def _retrieval_documents(
...
+            DOCUMENT_METADATA: document.get("metadata") or {},
         }
         for document in (run.get("outputs") or {}).get("documents") or []
     ]
 
 
 class OpenInferenceTracer(Tracer, BaseTracer):
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self._exporter = self._exporter or HttpExporter()
```
Comment (mikeldking) [inline src/phoenix/experimental/callbacks/langchain_tracer.py:168]: Should we at least add the option to swap this out as a constructor kwarg?

File: src/phoenix/experimental/callbacks/langchain_tracer.py [hunk 17]
```diff
@@ -7,6 +7,7 @@
 from langchain.callbacks.tracers.base import BaseTracer
 from langchain.callbacks.tracers.schemas import Run
 
+from phoenix.trace.exporter import HttpExporter
```
Comment (mikeldking) [inline src/phoenix/experimental/callbacks/langchain_tracer.py:10]: Isn't this really an HTTProtoExporter?

File: src/phoenix/experimental/callbacks/langchain_tracer.py (hunk 16 above)
Comment (mikeldking) [reply src/phoenix/experimental/callbacks/langchain_tracer.py:168]: Also could let it be None...

File: src/phoenix/proto/trace/v1/trace.proto [hunk 18]
```diff
@@ -0,0 +1,83 @@
+syntax = "proto3";
```
Comment (mikeldking) [inline src/phoenix/proto/trace/v1/trace.proto:1]: this is nested under `v1` - are we truly intending to support multiple versions? Because that might be painful to do.

File: src/phoenix/server/api/input_types/SpanSort.py [hunk 19]
```diff
@@ -1,33 +1,62 @@
...
     LLM_TOKEN_COUNT_COMPLETION,
     LLM_TOKEN_COUNT_PROMPT,
     LLM_TOKEN_COUNT_TOTAL,
-    START_TIME,
 )
-from phoenix.server.api.types.SortDir import SortDir
+
+
+def _get_attribute_value(span: Span, key: str) -> Any:
+    return span.attributes.get(key)
```
Comment (mikeldking) [inline src/phoenix/server/api/input_types/SpanSort.py:25]: why not just put this on the Span class?

File: src/phoenix/server/api/input_types/SpanSort.py [hunk 20]
```diff
@@ -1,33 +1,62 @@
...
+        key=CUMULATIVE_LLM_TOKEN_COUNT_TOTAL,
+    )
+    cumulativeTokenCountPrompt = partial(
+        _get_attribute_value,
+        key=CUMULATIVE_LLM_TOKEN_COUNT_PROMPT,
+    )
+    cumulativeTokenCountCompletion = partial(
+        _get_attribute_value,
+        key=CUMULATIVE_LLM_TOKEN_COUNT_COMPLETION,
+    )
```
Comment (mikeldking) [inline src/phoenix/server/api/input_types/SpanSort.py:59]: Using the enum value as the getter lambda feels a bit like a strange DSL that - while makes the code look brief, it feels like a level of brevity and indirection that ultimately is kinda more complex than needed. I'd consider leaving the enum and enum and just have a function that uses the enum as a parameter. That will make the code easier to test, follow, and debug.

File: src/phoenix/server/api/schema.py [hunk 21]
```diff
@@ -10,19 +11,20 @@
 from strawberry.types import Info
 from typing_extensions import Annotated
 
-from phoenix.datetime_utils import floor_to_minute
+import phoenix.trace.schemas as s
```
Comment (mikeldking) [inline src/phoenix/server/api/schema.py:14]: not crazy about the `s` abbreviation. Can we name it in plain english

File: src/phoenix/server/api/schema.py [hunk 22]
```diff
@@ -203,44 +204,39 @@ def hdbscan_clustering(
...
-                    axis=1,
+                chain.from_iterable(
+                    map(traces.get_trace, map(UUID, trace_ids)),
+                )
+                if trace_ids
+                else traces.get_spans(
+                    start_time=time_range.start if time_range else None,
+                    stop_time=time_range.end if time_range else None,
+                    root_spans_only=root_spans_only,
                 )
```
Comment (mikeldking) [inline src/phoenix/server/api/schema.py:232]: the level of conditionals and nesting here makes this kinda hard to read - I'd consider just using simpler syntax to make it a bit readable.

File: src/phoenix/server/api/types/Span.py [hunk 23]
```diff
@@ -1,36 +1,25 @@
...
```
Comment (mikeldking) [inline src/phoenix/server/api/types/Span.py:11]:
```suggestion
import phoenix.trace.schemas as trace_schema
```

File: src/phoenix/server/app.py [hunk 24]
```diff
@@ -104,6 +109,37 @@ async def get(self, request: Request) -> FileResponse:
...
 
+T = TypeVar("T", contravariant=True)
+
+
+class SupportsPut(Protocol[T]):
+    def put(self, obj: T) -> None:
+        ...
+
+
+class SpanHandler(HTTPEndpoint):
```
Comment (mikeldking) [inline src/phoenix/server/app.py:120]: let's move this into a separate file so the app entrypoint is a bit cleaner.

File: src/phoenix/server/app.py [hunk 25]
```diff
@@ -17,10 +18,14 @@
...
-from phoenix.core.model_schema import Model
-from phoenix.core.traces import Traces
+import phoenix.trace.v1.trace_pb2 as pb
 
+from ..config import SERVER_DIR
+from ..core.model_schema import Model
+from ..core.traces import Traces
+from ..trace.schemas import Span
+from ..trace.span_json_decoder import json_to_span
+from ..trace.v1 import encode
```
Comment (mikeldking) [inline src/phoenix/server/app.py:28]: I'd convert these to non-relative imports

File: src/phoenix/server/app.py [hunk 26]
```diff
@@ -104,6 +109,37 @@ async def get(self, request: Request) -> FileResponse:
         )
 
 
+T = TypeVar("T", contravariant=True)
+
+
+class SupportsPut(Protocol[T]):
+    def put(self, obj: T) -> None:
+        ...
```
Comment (mikeldking) [inline src/phoenix/server/app.py:117]: The generic for the one case seems strange. Why not explicitly type it as the queue?

File: src/phoenix/server/main.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mikeldking) [reply src/phoenix/server/main.py:121]: but if you use `.put` in the constructor I do agree it would be cleaner.

File: src/phoenix/trace/exporter.py [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+        )
+        self._queue: "SimpleQueue[Optional[Span]]" = SimpleQueue()
+        # Putting `None` as the sentinel value for queue termination.
+        weakref.finalize(self, self._queue.put, None)
+        Thread(target=self._consume_spans, daemon=True).start()
+
+    def export(self, span: Span) -> None:
+        self._queue.put(span)
+
+    def _consume_spans(self) -> None:
```
Comment (mikeldking) [inline src/phoenix/trace/exporter.py:42]: the name `consume` is kinda confusing here, especially in the context of an exporter. Something like `process_spans` would be better because `consume` has meaning in the producer consumer pattern.

File: src/phoenix/trace/filter.py [hunk 28]
```diff
@@ -0,0 +1,180 @@
...
+            if variable_name.isupper()
+            and variable_name.startswith(("RETRIEVAL", "EMBEDDING", "LLM", "TOOL"))
+        )
+    }.items():
+        yield source_segment, ast_replacement
+        yield "attributes." + source_segment, ast_replacement
+        yield "span.attributes." + source_segment, ast_replacement
+
+
+class _Translator(ast.NodeTransformer):
```
Comment (mikeldking) [inline src/phoenix/trace/filter.py:86]: Keeping this private via _ just makes the code just harder to read - I'm not sure it's worth it

File: src/phoenix/trace/filter.py [hunk 29]
```diff
@@ -0,0 +1,180 @@
...
+        return cast(bool, eval(self._compiled, {}, {"span": span, "_MISSING": _Missing()}))
+
+
+def _replace_none_with_missing(
+    value: ast.expr,
+    as_str: bool = False,
+) -> ast.IfExp:
+    """E.g. `value` becomes
+    `_MISSING if (_MAYBE := value) is None else _MAYBE`
+    """
```
Comment (mikeldking) [inline src/phoenix/trace/filter.py:26]:
```suggestion
    """
    E.g. `value` becomes
    `_MISSING if (_MAYBE := value) is None else _MAYBE`
    """
```

File: src/phoenix/trace/filter.py [hunk 30]
```diff
@@ -0,0 +1,180 @@
+import ast
+from typing import Any, Iterator, Mapping, Tuple, cast
+
+from phoenix.trace import semantic_conventions
+from phoenix.trace.schemas import Span
+
+
+class Filter:
```
Comment (mikeldking) [inline src/phoenix/trace/filter.py:8]:
```suggestion
class SpanFilter:
```

File: src/phoenix/trace/filter.py (hunk 29 above)
Comment (mikeldking) [reply src/phoenix/trace/filter.py:26]: I'm not sure I follow this function definition. Can you add more examples? I don't know what `_MAYBE` is

File: src/phoenix/trace/filter.py [hunk 31]
```diff
@@ -0,0 +1,180 @@
...
+                ast.operator,
+                ast.unaryop,
+            ),
+        ):
+            continue
+        source_segment: str = cast(str, ast.get_source_segment(source, node))
+        raise SyntaxError(f"invalid expression: {source_segment}")  # TODO: add details
+
+
+class _Missing:
```
Comment (mikeldking) [inline src/phoenix/trace/filter.py:143]: These _ prefixes are really hindering readability - I don't really feel it's worth it.

File: src/phoenix/session/session.py [hunk 32]
```diff
@@ -133,6 +140,28 @@ def url(self) -> str:
...
         return _get_url(self.port, self.is_colab)
 
+    def get_spans(
+        self,
+        filter_condition: Optional[str] = None,
+        *,
+        start_time: Optional[datetime] = None,
+        stop_time: Optional[datetime] = None,
+        root_spans_only: Optional[bool] = None,
+    ) -> Optional[pd.DataFrame]:
```
Comment (axiomofjoy) [inline src/phoenix/session/session.py:150]: Since we have a `Span` type, we might want to reserve `get_spans`. What do you think of `get_span_dataframe`?

File: src/phoenix/server/app.py [hunk 33]
```diff
@@ -104,6 +109,37 @@ async def get(self, request: Request) -> FileResponse:
...
+        try:
+            content_type = request.headers.get("content-type")
+            if content_type == "application/x-protobuf":
+                body = await request.body()
+                content_encoding = request.headers.get("content-encoding")
+                if content_encoding == "gzip":
+                    body = gzip.decompress(body)
+                pb_span = pb.Span()
+                pb_span.ParseFromString(body)
+            else:
```
Comment (axiomofjoy) [inline src/phoenix/server/app.py:133]: Might be worth an explicit check for "application/json", especially since this endpoint returns a 422 on error.

File: src/phoenix/server/app.py [hunk 34]
```diff
@@ -104,6 +109,37 @@ async def get(self, request: Request) -> FileResponse:
...
+                span = json_to_span(await request.json())
+                assert isinstance(span, Span)
```
Comment (axiomofjoy) [inline src/phoenix/server/app.py:135]: It's strange to me that `json_to_span` can return anything and that a type check is needed here.

**Comment by axiomofjoy** [comment]
Nice job on this one @RogerHYang, I learned a lot reading through it.

File: src/phoenix/core/traces.py [hunk 35]
```diff
@@ -1,108 +1,267 @@
...
-        parent_span_ids = df.loc[:, PARENT_ID]
-        for column, cumulative_column in (
+    def __init__(self, spans: Optional[Iterable[Span]] = None) -> None:
+        self._queue: "SimpleQueue[Optional[pb.Span]]" = SimpleQueue()
+        # Putting `None` as the sentinel value for queue termination.
+        weakref.finalize(self, self._queue.put, None)
+        for span in spans or ():
+            self.put(span)
+        self._lock = RLock()
+        self._spans: Dict[SpanID, ReadableSpan] = {}
```
[the remark this answers is not in the document]
Comment (mikeldking) [reply src/phoenix/core/traces.py:110]: I think I lean non-proto, proto encoding is meant for transport but I think having to make things conform to proto for normal cases is non-ideal overhead.

File: src/phoenix/core/traces.py [hunk 36]
```diff
@@ -1,108 +1,267 @@
...
+        )
+        self._start_time_sorted_root_span_ids: SortedKeyList[SpanID] = SortedKeyList(
+            key=lambda span_id: self._spans[span_id].start_time.ToDatetime(timezone.utc),
+        )
+        self._latency_sorted_root_span_ids: SortedKeyList[SpanID] = SortedKeyList(
+            key=lambda span_id: self._spans[span_id][LATENCY_MS],
+        )
+        self._min_start_time: Optional[datetime] = None
+        self._max_start_time: Optional[datetime] = None
+        self._start_consumer()
```
[the remark this answers is not in the document]
Comment (mikeldking) [reply src/phoenix/core/traces.py:126]: We can leave it for now but I do still have this concern where transport code ends up proliferating business logic. It's something that can cause a cascade of strange dependancies.

I agree the encapsulation of a stream makes sense but there are benefits to having derived streams that are observable on their own merit. 

Let's see how it goes.

[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/core/traces.py:126]: ok. so it seems to me that having it decoded at te receiver (i.e. at the `v1/spans` endpoint) would resolve your concern because Traces now no longer knows about protos or anything from the transportation layer
